File: lib/js/belt_MutableQueue.js
'use strict';

var Curry = require("./curry.js");

var $$null = null;

function make() {
  return {
          length: 0,
          first: $$null,
          last: $$null
        };
}

function clear(q) {
  q.length = 0;
  q.first = $$null;
  q.last = $$null;
  return /* () */0;
}

function add(q, x) {
  var cell = {
    content: x,
    next: $$null
  };
  var match = q.last;
  if (match !== null) {
    q.length = q.length + 1 | 0;
    match.next = cell;
    q.last = cell;
    return /* () */0;
  } else {
    q.length = 1;
    q.first = cell;
    q.last = cell;
    return /* () */0;
  }
}

function peek(q) {
  var match = q.first;
  if (match !== null) {
    return /* Some */[match.content];
  } else {
    return /* None */0;
  }
}

function peekUndefined(q) {
  var match = q.first;
  if (match !== null) {
    return match.content;
  } else {
    return undefined;
  }
}

function peekExn(q) {
  var match = q.first;
  if (match !== null) {
    return match.content;
  } else {
    throw new Error("Belt.Queue.Empty");
  }
}

function pop(q) {
  var match = q.first;
  if (match !== null) {
    var next = match.next;
    if (next === null) {
      clear(q);
      return /* Some */[match.content];
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return /* Some */[match.content];
    }
  } else {
    return /* None */0;
  }
}

function popExn(q) {
  var match = q.first;
  if (match !== null) {
    var next = match.next;
    if (next === null) {
      clear(q);
      return match.content;
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return match.content;
    }
  } else {
    throw new Error("Empty");
  }
}

function popUndefined(q) {
  var match = q.first;
  if (match !== null) {
    var next = match.next;
    if (next === null) {
      clear(q);
      return match.content;
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return match.content;
    }
  } else {
    return undefined;
  }
}

function copy(q) {
  var qRes = {
    length: q.length,
    first: $$null,
    last: $$null
  };
  var _prev = $$null;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var prev = _prev;
    if (cell !== null) {
      var content = cell.content;
      var res = {
        content: content,
        next: $$null
      };
      if (prev !== null) {
        prev.next = res;
      } else {
        qRes.first = res;
      }
      _cell = cell.next;
      _prev = res;
      continue ;
    } else {
      qRes.last = prev;
      return qRes;
    }
  };
}

function mapU(q, f) {
  var qRes = {
    length: q.length,
    first: $$null,
    last: $$null
  };
  var _prev = $$null;
  var _cell = q.first;
  var f$1 = f;
  while(true) {
    var cell = _cell;
    var prev = _prev;
    if (cell !== null) {
      var content = f$1(cell.content);
      var res = {
        content: content,
        next: $$null
      };
      if (prev !== null) {
        prev.next = res;
      } else {
        qRes.first = res;
      }
      _cell = cell.next;
      _prev = res;
      continue ;
    } else {
      qRes.last = prev;
      return qRes;
    }
  };
}

function map(q, f) {
  return mapU(q, Curry.__1(f));
}

function isEmpty(q) {
  return q.length === 0;
}

function size(q) {
  return q.length;
}

function forEachU(q, f) {
  var _cell = q.first;
  var f$1 = f;
  while(true) {
    var cell = _cell;
    if (cell !== null) {
      f$1(cell.content);
      _cell = cell.next;
      continue ;
    } else {
      return /* () */0;
    }
  };
}

function forEach(q, f) {
  return forEachU(q, Curry.__1(f));
}

function reduceU(q, accu, f) {
  var f$1 = f;
  var _accu = accu;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var accu$1 = _accu;
    if (cell !== null) {
      var accu$2 = f$1(accu$1, cell.content);
      _cell = cell.next;
      _accu = accu$2;
      continue ;
    } else {
      return accu$1;
    }
  };
}

function reduce(q, accu, f) {
  return reduceU(q, accu, Curry.__2(f));
}

function transfer(q1, q2) {
  if (q1.length > 0) {
    var match = q2.last;
    if (match !== null) {
      q2.length = q2.length + q1.length | 0;
      match.next = q1.first;
      q2.last = q1.last;
      return clear(q1);
    } else {
      q2.length = q1.length;
      q2.first = q1.first;
      q2.last = q1.last;
      return clear(q1);
    }
  } else {
    return 0;
  }
}

function fillAux(_i, arr, _cell) {
  while(true) {
    var cell = _cell;
    var i = _i;
    if (cell !== null) {
      arr[i] = cell.content;
      _cell = cell.next;
      _i = i + 1 | 0;
      continue ;
    } else {
      return /* () */0;
    }
  };
}

function toArray(x) {
  var v = new Array(x.length);
  fillAux(0, v, x.first);
  return v;
}

function fromArray(arr) {
  var q = make(/* () */0);
  for(var i = 0 ,i_finish = arr.length - 1 | 0; i <= i_finish; ++i){
    add(q, arr[i]);
  }
  return q;
}

exports.make = make;
exports.clear = clear;
exports.isEmpty = isEmpty;
exports.fromArray = fromArray;
exports.add = add;
exports.peek = peek;
exports.peekUndefined = peekUndefined;
exports.peekExn = peekExn;
exports.pop = pop;
exports.popUndefined = popUndefined;
exports.popExn = popExn;
exports.copy = copy;
exports.size = size;
exports.mapU = mapU;
exports.map = map;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.reduceU = reduceU;
exports.reduce = reduce;
exports.transfer = transfer;
exports.toArray = toArray;
/* No side effect */

File: lib/ocaml/belt_MutableQueue.ml
(**************************************************************************)
(*                                                                        *)
(*                                 OCaml                                  *)
(*                                                                        *)
(*         Francois Pottier, projet Cristal, INRIA Rocquencourt           *)
(*                  Jeremie Dimino, Jane Street Europe                    *)
(*                                                                        *)
(*   Copyright 2002 Institut National de Recherche en Informatique et     *)
(*     en Automatique.                                                    *)
(*                                                                        *)
(*   All rights reserved.  This file is distributed under the terms of    *)
(*   the GNU Lesser General Public License version 2.1, with the          *)
(*   special exception on linking described in the file LICENSE.          *)
(*                                                                        *)
(**************************************************************************)
(* Adapted significantly by BuckleScript Authors                          *)
module A = Belt_Array

type 'a node = { 
  content: 'a; 
  mutable next: 'a cell 
}
and 'a cell = 'a node Js.null
and 'a t = {
  mutable length: int;
  mutable first: 'a cell;
  mutable last: 'a cell
} [@@bs.deriving abstract]

let null  = Js.null 
let return = Js_null.return 

let make () = 
  t 
    ~length: 0
    ~first:null
    ~last:null


let clear q =
  lengthSet q  0;
  firstSet q  null;
  lastSet q  null

let add q x =
  let cell = return @@ node 
      ~content:x
      ~next:null in
  match Js.nullToOption (lastGet q )with
  | None ->
    lengthSet  q 1;
    firstSet q cell;
    lastSet q cell
  | Some last ->
    lengthSet q (lengthGet q + 1);
    nextSet last  cell;
    lastSet q  cell


  
let peek q =
  match Js.nullToOption (firstGet q ) with
  | None -> None
  | Some v -> Some (contentGet v)

let peekUndefined q =
  match Js.nullToOption (firstGet q ) with
  | None -> Js.undefined
  | Some v -> Js.Undefined.return (contentGet v)


let peekExn q =
  match Js.nullToOption (firstGet q ) with
  | None -> [%assert "Belt.Queue.Empty"]
  | Some v -> contentGet v

let pop q =
  match Js.nullToOption (firstGet q ) with
  | None -> None
  | Some x  ->
    let next = nextGet x in 
    if next = Js.null then 
      begin (* only one element*)
        clear q;
        Some (contentGet x)
      end
    else begin 
      lengthSet q (lengthGet q - 1);
      firstSet q next;
      Some(contentGet x) 
    end

let popExn q =
  match Js.nullToOption (firstGet q ) with
  | None -> [%assert "Empty"]
  | Some x  ->
    let next = nextGet x in 
    if next = Js.null then 
      begin (* only one element*)
        clear q;
        contentGet x
      end
    else begin 
      lengthSet q (lengthGet q - 1);
      firstSet q next;
      contentGet x 
    end    

let popUndefined q =
  match Js.nullToOption (firstGet q ) with
  | None -> Js.undefined
  | Some x  ->
    let next = nextGet x in 
    if next = Js.null then 
      begin (* only one element*)
        clear q;
        Js.Undefined.return (contentGet x)
      end
    else begin 
      lengthSet q (lengthGet q - 1);
      firstSet q next;
      Js.Undefined.return (contentGet x) 
    end

let rec copyAux qRes prev cell =
  match Js.nullToOption cell with
  | None -> lastSet qRes  prev; qRes
  | Some x  ->
    let content = contentGet x in 
    let res = return @@ node ~content ~next:null in
    begin match Js.nullToOption prev with
      | None -> firstSet qRes res
      | Some p -> nextSet p  res
    end;
    copyAux qRes res (nextGet x)

let copy q =
  copyAux (t  ~length:(lengthGet q) ~first:null ~last:null)  null (firstGet q)


      
let rec copyMapAux qRes prev cell f =
  match Js.nullToOption cell with
  | None -> lastSet qRes  prev; qRes
  | Some x  ->
    let content = f (contentGet x) [@bs] in 
    let res = return @@ node ~content ~next:null in
    begin match Js.nullToOption prev with (*TODO: optimize to remove such check*)
      | None -> firstSet qRes res
      | Some p -> nextSet p  res
    end;
    copyMapAux qRes res (nextGet x) f

let mapU q f =
  copyMapAux (t  ~length:(lengthGet q) ~first:null ~last:null)  null (firstGet q) f
    
let map q f = mapU q (fun [@bs] a -> f a)
    
let isEmpty q =
  lengthGet q = 0

let size q =
  lengthGet q

let rec iterAux cell f =
  match Js.nullToOption cell with
  | None -> ()
  | Some x  ->
    f (contentGet x) [@bs];
    iterAux (nextGet x) f

let forEachU q f =
  iterAux (firstGet q) f

let forEach q f = forEachU q (fun[@bs] a -> f a)
    
let rec foldAux f accu cell =
  match Js.nullToOption cell with
  | None -> accu
  | Some x  ->
    let accu = f accu (contentGet x) [@bs] in
    foldAux f accu (nextGet x)

let reduceU q  accu f =
  foldAux f accu (firstGet q)

let reduce q accu f = reduceU q accu (fun[@bs] a b -> f a b)
    
let transfer q1 q2 =
  if lengthGet q1 > 0 then
    match Js.nullToOption (lastGet q2) with
    | None ->
      lengthSet q2 (lengthGet q1);
      firstSet q2 (firstGet q1);
      lastSet q2 (lastGet q1);
      clear q1
    | Some l ->
      lengthSet q2  (lengthGet q2 + lengthGet q1);
      nextSet l (firstGet q1);
      lastSet q2 (lastGet  q1);
      clear q1

let rec fillAux i arr cell =       
  match Js.nullToOption cell with 
  | None -> ()
  | Some x ->
    A.setUnsafe arr i (contentGet x);
    fillAux (i + 1) arr (nextGet x) 

let toArray x =         
  let v = A.makeUninitializedUnsafe (lengthGet x) in 
  fillAux 0 v (firstGet x);
  v

(*TODO: optimize *)
let fromArray arr =
  let q = make () in
  for i = 0 to A.length arr - 1 do
    add q (A.getUnsafe arr i)
  done ;
  q

